src/rules/no_duplicate_case.rs

```diff
@@ -128,9 +188,172 @@ switch (someText) {
+        {
+          col: 102,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        }
+      ],
+      "var a = 1, f = function(s) { return { p1: s } }; switch (a) {case f(a + 1).p1: break; case f(a + 1).p1: break; default: break;}": [
+        {
+          col: 86,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        }
+      ],
+      "var a = 1, f = function(s) { return { p1: s } }; switch (a) {case f(a === 1 ? 2 : 3).p1: break; case f(a === 1 ? 2 : 3).p1: break; default: break;}": [
+        {
+          col: 96,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        }
+      ],
+      "var a = 1, f1 = function() { return { p1: 1 } }; switch (a) {case f1().p1: break; case f1().p1: break; default: break;}": [
+        {
+          col: 82,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        }
+      ],
+      "var a = [1, 2]; switch(a.toString()){case ([1, 2]).toString():break; case ([1, 2]).toString():break; default:break;}": [
+        {
+          col: 69,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        }
+      ],
+      "switch (a) { case a: case a: }": [
+        {
+          col: 21,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        }
+      ],
+      "switch (a) { case a: break; case b: break; case a: break; case c: break; case a: break; }": [
+        {
+          col: 43,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        },
+        {
+          col: 73,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        }
+      ],
+      "var a = 1, f = function(s) { return { p1: s } }; switch (a) {case f(a + 1).p1: break; case f(a+1).p1: break; default: break;}": [
+        {
+          col: 86,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        }
+      ],
+
+      // nested
+      r#"
+switch (a) {
+  case 1:
+    switch (b) {
+      case 1:
+        break;
+      case 1:
+        break;
+    }
+  default:
+    break;
+}
+      "#: [
+        {
+          line: 7,
+          col: 6,
+          message: NoDuplicateCaseMessage::Unexpected,
+          hint: NoDuplicateCaseHint::RemoveOrRename,
+        }
+      ],
+
+      // TODO(magurotuna): to pass the following tests, we have to remove comments and whitespaces somehow.
+      // "var a = 1, p = {p: {p1: 1, p2: 1}}; switch (a) {case p.p.p1: break; case p. p // comment\n .p1: break; default: break;}": [
```

Comment:
@magurotuna can you elaborate on this one?

Reply:
@bartlomieju 
I want to do so, but I don't come up with an idea yet.
It seems like we should implement something to remove comments by ourselves, which I don't think is a good way.

@kdy1 
Can I ask for your help?
I'd like to check if there are duplicates in test sections of a switch statement. In the current implementation, it's done by comparing  two tests in text representation, like this:

```rust
let test_txt = source_map
  .span_to_snippet(test_span)
  .unwrap()
  .replace(|c: char| c.is_whitespace(), "");

if seen.contains(&test_txt) {
  // there is duplicate
}
```

However, this way we cannot handle cases where tests have any comment inside. e.g:

```typescript
// we want the following snippet to be reported

switch (a) {
  case p.p1:
    break;
  case p. /* some comment */ p1:
    break;
}
```

So I think removing comments is straight forward approach here.
If you come up with any better solution, I'd be happy to hear it.

Reply:
You can clone the discriminator of switch statement and remove span and then compare. It's fast enough.

There's a function named [`drop_span`](https://swc.rs/rustdoc/swc_ecma_utils/fn.drop_span.html) and [`DropSpan`](https://swc.rs/rustdoc/swc_ecma_utils/struct.DropSpan.html) in `swc_ecmascript::utils`.

Reply:
> There's a function named drop_span and DropSpan in swc_ecmascript::utils. 

Good to know, I think we have duplicate implementation in `src/swc_util.rs`

Reply:
Great, I'll try it, thanks for the quick reply.

Reply:
@bartlomieju 

Can I work on removing `DropSpan` from `src/swc_util.rs`?

Reply:
@magurotuna sure :)